File: AVLTree/cpp/AVLTree.cpp
#include "AVLTree.h"
#include <iostream>

AVLTree::AVLTree ():
    root    (nullptr)
{}

void AVLTree::destruct (Node* _node)
{
    if (_node == nullptr)
        return;

    destruct(_node->left);
    destruct(_node->right);

    delete _node;
}

AVLTree::~AVLTree ()
{
    destruct(root);
}

Node* AVLTree::do_insert (Node* _node, int _key)
{
    if (_node == nullptr)
        return new Node(_key);

    if (_key < _node->key)
        _node->left = avl_insert(_node->left, _key);

    else if (_key > _node->key)
        _node->right = avl_insert(_node->right, _key);

    update_height(_node);

    int balance_factor = get_balance_factor(_node);

    if (balance_factor < -1)
    {
        // Случай LL
        if (_key < _node->left->key)
        {
            _node = rotate_right(_node);
        }
        // Случай LR
        else if (_key > _node->left->key)
        {
            _node->left = rotate_left(_node->left);           
            _node = rotate_right(_node);
        }
    }
    else if (balance_factor > 1)
    {
        // Случай RR
        if (_key > _node->right->key)
        {
            _node = rotate_left(_node);
        }
        // Случай RL
        else if (_key < _node->right->key)
        {
            _node->right = rotate_right(_node->right);
            _node = rotate_left(_node);
        }
    }

    return _node;
}

void AVLTree::insert (int _key)
{
    root = do_insert(root, _key);
}

void AVLTree::do_print_in_order (Node* _node)
{
    if (_node == nullptr)
        return;

    do_print_in_order(_node->left);

    std::cout << _node->key << " ";

    do_print_in_order(_node->right);
}

void AVLTree::print_in_order ()
{
    do_print_in_order(root);
    std::cout << "\n" << std::endl;
}

bool AVLTree::contains (int _key)
{
    return find(_key) != nullptr;
}

Node* AVLTree::do_find (Node* _node, int _key)
{
    if (_node == nullptr)
        return nullptr;

    if (_key == _node->key)
        return _node;
    
    else if (_key < _node->key)
        return do_find(_node->left, _key);
    
    else
        return do_find(_node->right, _key);
}

Node* AVLTree::find (int _key)
{
    return do_find(root, _key);
}

File: AVLTree/py/AVLTree.py
from Node import Node

class AVLTree:
    def __init__(self):
        self.root = None

    # def destruct(self, node):
    #     if node is None:
    #         return
    #
    #     self.destruct(node.left)
    #     self.destruct(node.right)
    #
    #     del node

    # def __del__(self):
    #     self.destruct(self.root)

    def update_height(self, node):
        if node is None:
            return 0

        left_height = self.update_height(node.left) if node.left else 0
        right_height = self.update_height(node.right) if node.right else 0

        node.height = max(left_height, right_height) + 1
        return node.height

    @staticmethod
    def get_balance_factor(node):
        if node is None:
            return 0

        left_height = node.left.height if node.left else 0
        right_height = node.right.height if node.right else 0

        return right_height - left_height

    def rotate_right(self, node):
        new_root = node.left
        node.left = new_root.right
        new_root.right = node

        self.update_height(node)
        self.update_height(new_root)

        return new_root

    def rotate_left(self, node):
        new_root = node.right
        node.right = new_root.left
        new_root.left = node

        self.update_height(node)
        self.update_height(new_root)

        return new_root

    def do_insert(self, node, key):
        if node is None:
            return Node(key)

        if key < node.key:
            node.left = self.do_insert(node.left, key)
        elif key > node.key:
            node.right = self.do_insert(node.right, key)

        self.update_height(node)

        balance_factor = self.get_balance_factor(node)

        if balance_factor < -1:
            # Случай LL
            if key < node.left.key:
                node = self.rotate_right(node)
            # Случай LR
            elif key > node.left.key:
                node.left = self.rotate_left(node.left)
                node = self.rotate_right(node)
        elif balance_factor > 1:
            # Случай RR
            if key > node.right.key:
                node = self.rotate_left(node)
            # Случай RL
            elif key < node.right.key:
                node.right = self.rotate_right(node.right)
                node = self.rotate_left(node)

        return node

    def insert(self, key):
        self.root = self.do_insert(self.root, key)

    def do_print_in_order(self, node):
        if node is None:
            return

        self.do_print_in_order(node.left)

        print(node.key, end=" ")

        self.do_print_in_order(node.right)

    def print_in_order(self):
        self.do_print_in_order(self.root)
        print("\n")

    def contains(self, key):
        return self.find(key) is not None

    def do_find(self, node, key):
        if node is None:
            return None

        if key == node.key:
            return node

        elif key < node.key:
            return self.do_find(node.left, key)

        else:
            return self.do_find(node.right, key)

    def find(self, key):
        return self.do_find(self.root, key)